typeset/paras.js

```diff
@@ -0,0 +1,105 @@
+/*
+1.
+2. If there is a multiline replacement and more replacements on this line
+3. If there is a multiline replacement and no additional replacements on this line
+4. If there is a multiline replacement and no additional replacements
+5. If there are no additional replacements
+6. If there are multiple replacements on this line
+7. If the current replacement extends past this line
+8. If there are more replacements, but none on this line
+*/
+
+function expectValue (v, msg = 'Expected a value but got nothing') {
+  if (v == null) {
+    throw new Error(msg)
+  }
+  return v
+}
+
+// Pipe And Replace Arbitrary Stuff
+function PARAS (replacements, input, output) {
+  let lineNumber = 0
+  const spillover = []
+  let replaceTo
+  let finished = false
+
+  function handleLine (line) {
+    let lastEndCol = 0
+    // If the line started in the last chunk, concat here!
+    if (spillover.length) {
+      line = spillover.join('') + line
+      spillover.splice(0)
+    }
+    if (finished) {
+      output.write(line)
+      return
+    }
+    // Handle multiline replacements
+    if (replaceTo !== undefined) {
+      if (lineNumber !== replaceTo[0]) return
+      lastEndCol = replaceTo[1] - 1
+      replaceTo = undefined
+      // If there are no more replacements on this line
+      if (replacements.length === 0 || replacements[0].posStart[0] !== lineNumber) {
+        output.write(line.slice(lastEndCol))
+        finished = replacements.length === 0
+        return
+      }
+    } else if (replacements.length === 0) {
+      output.write(line)
+      finished = true
+      return
+    }
+    if (replacements[0].posStart[0] === lineNumber) {
+      while (true) {
+        const replacement = replacements.shift()
+        output.write(line.slice(lastEndCol, replacement.posStart[1] - 1))
+        output.write(expectValue(replacement.substitution))
+        // We cannot do any more now
+        if (replacement.posEnd[0] > lineNumber) {
+          replaceTo = replacement.posEnd
+          // When there are replacements that fall between this one and its end line,
+          // they are implicitly replaced. Just remove them from the list.
```

Comment:
If you have xml like 
```xml
<a>
    <a>some text</a>
</a>
```
and you want to replace `a` tags, the nested tag will match too. This while loop was specifically for that case. Granted, I do not think it is valid to have nested math tags in mathml, so this situation probably will not ever happen in mathify.

Reply:
But is there a way for the inner `<a>` replacement to show up in the resulting document? It seems like the inner `<a>` is always discarded since there is no way for a replacement to define a "slot" for nested replacements to go to... if that's the case then it seems like this code could error/warn and the scan-xml `recorders` could be simplified to be an optional single `recorder`, no?